src/frontend/src/handler/alter_table_column.rs

```diff
@@ -52,71 +49,47 @@ pub async fn get_new_table_definition_for_cdc_table(
 ) -> Result<(Statement, Arc<TableCatalog>)> {
     let original_catalog = fetch_table_catalog_for_alter(session.as_ref(), &table_name)?;
 
-    // Retrieve the original table definition and parse it to AST.
-    let mut definition = original_catalog.create_sql_ast()?;
-
-    let Statement::CreateTable {
-        columns: original_columns,
-        format_encode,
-        constraints,
-        ..
-    } = &mut definition
-    else {
-        panic!("unexpected statement: {:?}", definition);
-    };
-
-    assert!(
-        format_encode.is_none(),
-        "source schema should be None for CDC table"
+    assert_eq!(
+        original_catalog.row_id_index, None,
+        "primary key of cdc table must be user defined"
     );
 
-    if bind_table_constraints(constraints)?.is_empty() {
-        // For table created by `create table t (*)` the constraint is empty, we need to
-        // retrieve primary key names from original table catalog if available
-        let pk_names: Vec<_> = original_catalog
-            .pk
-            .iter()
-            .map(|x| original_catalog.columns[x.column_index].name().to_owned())
-            .collect();
-
-        constraints.push(TableConstraint::Unique {
-            name: None,
-            columns: pk_names.iter().map(Ident::new_unchecked).collect(),
-            is_primary: true,
-        });
-    }
+    // Retrieve the original table definition.
+    let mut definition = original_catalog.create_sql_ast()?;
 
-    let orig_column_catalog: HashMap<String, ColumnCatalog> = HashMap::from_iter(
-        original_catalog
-            .columns()
-            .iter()
-            .map(|col| (col.name().to_owned(), col.clone())),
-    );
+    // Clear the original columns field, so that we'll follow `new_columns` to generate a
+    // purified definition.
+    {
+        let Statement::CreateTable {
+            columns,
+            constraints,
+            ..
+        } = &mut definition
+        else {
+            panic!("unexpected statement: {:?}", definition);
+        };
 
-    // update the original columns with new version columns
-    let mut new_column_defs = vec![];
-    for new_col in new_columns {
-        // if the column exists in the original catalog, use it to construct the column definition.
-        // since we don't support altering the column type right now
-        if let Some(original_col) = orig_column_catalog.get(new_col.name()) {
-            let ty = original_col.data_type().to_ast();
-            new_column_defs.push(ColumnDef::new(original_col.name().into(), ty, None, vec![]));
-        } else {
-            let ty = new_col.data_type().to_ast();
-            new_column_defs.push(ColumnDef::new(new_col.name().into(), ty, None, vec![]));
-        }
+        columns.clear();
+        constraints.clear();
```

Comment:
The constraint is also cleared here, does `try_purify_table_source_create_sql_ast` can rebuild the constraint from the `new_columns`?

Reply:
Yes, for the primary key table constraint.